### PR TITLE
Feature/Issue 1142 OpenCL rectangular matrix multiply optimization

### PR DESCRIPTION
**Pull request by rok-cesnovar**
## Summary
This PR, together with @t4c1's vector * matrix multiplication PR, will finalize the matrix multiplication optimizations for now. Further optimizations are off course possible, but would mostly rely on techniques that require more careful tuning of parameters for the target GPU.

This PR speeds up matrix multiplication [N, K] x [K, M], where K>>N,M and NxM is not large 
(NxM <= 2,000,000). In these cases the number of created threads is small (number of threads = NxM) and we are thus not occupying all the compute cores and streaming multiprocessors on a GPU 
(occupancy is smaller for more high-end GPUs).

What this PR does is that it splits the scalar product between 2 or more threads. For instance [25,50] x [50, 25] would be split in 2 matrix multiplications of [25,25] x [25,25] and then adding the two resulting matrices together. For smaller size the speedups are nice, even for the cases of [200,5000] x [5000,200] we see a speedup of 1.5x. The below table shows the speedups for [N,K] x [K,N] where rows represent increasing N and columns increasing K. N=M in this case, the speedups for [100,5000] x [5000,100] or [1000,5000] x [5000,10], are the same. The speedups were measured on the Titan XP.

N\K | 5000 | 10000 | 20000 | 30000 | 40000 | 50000
-- | -- | -- | -- | -- | -- | --
25 | 5.648715241 | 7.213837355 | 8.869929932 | 10.19044284 | 11.69508974 | 12.47130435
50 | 3.988662438 | 4.856297549 | 6.086647454 | 7.852823031 | 8.438483646 | 8.485952363
75 | 3.488028731 | 3.575887986 | 3.356228723 | 3.691319806 | 3.98522295 | 3.877672985
100 | 2.180804785 | 2.291172491 | 2.28828944 | 2.226279633 | 2.435441739 | 2.397910012
150 | 1.695650065 | 1.655894943 | 1.809581678 | 1.747848855 | 1.640151875 | 1.671547052
200 | 1.467013117 | 1.510031276 | 1.506152636 | 1.498439427 | 1.511800047 | 1.493089874
300 | 1.314360821 | 1.339593042 | 1.366195018 | 1.392757012 | 1.37537513 | 1.407317773
400 | 1.564656409 | 1.550537634 | 1.695921757 | 1.819523503 | 1.894239896 | 1.900044516
500 | 1.228811176 | 1.228977324 | 1.255056579 | 1.257794444 | 1.31307825 | 1.311603841
600 | 1.280527408 | 1.32880363 | 1.302920378 | 1.372158873 | 1.326086226 | 1.39453303
700 | 1.24290718 | 1.271105308 | 1.394220753 | 1.403649005 | 1.375596479 | 1.338054425
800 | 1.210179125 | 1.250913979 | 1.248210011 | 1.316197441 | 1.327185358 | 1.292556195
900 | 1.122778925 | 1.119413461 | 1.127987443 | 1.129680365 | 1.123156502 | 1.120763962
1000 | 1.092608125 | 1.154006445 | 1.11553263 | 1.061182486 | 1.058325229 | 1.056288944
1500 | 1.052366383 | 1.027649973 | 1.013942153 | 1.03157486 | 1.040056041 | 1.021040896
2000 | 1.033078453 | 1.041143996 | 1.016128757 | 1.07237671 | 1.052827751 | 1.05E+00
3000 | 1.060574068 | 1.021190093 | 1.05E+00 | 1.02E+00 | 1.03E+00 | 1.03E+00
4000 | 1.000013069 | 1.006594855 | 1.07E+00 | 1.00E+00 | 1.00E+00 | 1.00E+00
5000 | 1.001837177 | 1.00E+00 | 1.01E+00 | 1.00E+00 | 1.00E+00 | 1.01E+00

As we can see that the speedups are gone once we hit larger N, since there are enough created threads to keep the GPU busy. The upper limit is set in the context to prevent splitting in those cases, since its of no use. The current limit is N*M <= 2,000,000. Not sure the speedups of a few % are worth the slightly larger footprint (these speedups are without the data transfers, so the end-to-end difference is smaller).

## Tests

Added tests to the `opencl/multiply`.

## Side Effects

There is a bit more of a memory footprint, but as this is limited to smaller size of NxM, that footprint is also limited. Worst case is 280MB for a multiply where the input matrices are 0.5GB+.

## Checklist

- [x] Math issue #1142 

- [x] Copyright holder:  Davor Sluga, Rok Češnovar (Univ. of Ljubljana)

- [x] the basic tests are passing

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by rok-cesnovar** [comment]
@serban-nicusor-toptal the release notes indicate that this has been merged, which it hasnt. Can the release notes be edited?

**Comment by serban-nicusor-toptal** [comment]
@rok-cesnovar updated! Thanks!

File: stan/math/opencl/kernels/add.hpp [hunk 1]
```diff
@@ -40,7 +40,43 @@ static const char *add_kernel_code = STRINGIFY(
  */
 const global_range_kernel<cl::Buffer, cl::Buffer, cl::Buffer, int, int> add(
     "add", {indexing_helpers, add_kernel_code});
+// \cond
+static const char *add_batch_kernel_code = STRINGIFY(
+    // \endcond
+    /**
+     * Adds all the matrices in a batch
```
Comment (seantalts) [inline stan/math/opencl/kernels/add.hpp]: Can you say a little more about this? Is this basically a sum function over a collection of matrices?

Comment (rok-cesnovar) [reply stan/math/opencl/kernels/add.hpp]: Yes, A contains a batch of K smaller matrices of size NxM. This kernel adds elements at matching indices into a resulting NxM matrix.

Comment (rok-cesnovar) [reply stan/math/opencl/kernels/add.hpp]: Added to doxygen.

**Pull request merged**